test/random_dem.cpp

```diff
@@ -496,6 +548,9 @@ int32_t random_dem_test(ptrdiff_t dims[2], uint32_t seed) {
   gwdt(dist, prev, costs, flats, heap, back, dims);
   test_gwdt(dist, prev, costs, flats, dims);
 
+  gradient8(gradient, dem, cellsize, unit, 0, dims);
```

Comment:
Should we also test the OpenMP version of this?

Reply:
Probably. One useful property-based test here would be that the single- and multithreaded version produce the same answers.

If I change this line to

```C
gradient8(gradient, dem, cellsize, unit, 1, dims);
```

the current test fails. Are you seeing the same thing?

Reply:
I think the test failure is because when we declare both loop variables before the OpenMP pragma, the loops end up sharing the inner `i` variable. Only the outer loops get divided between threads and then they race on the shared `i` variable.


See: https://stackoverflow.com/questions/75120618/openmp-data-races-without-a-collapse-clause

To fix this we either need to only predeclare the outer loop variable `j`:

```C
ptrdiff_t j;
#pragma omp parallel for if (use_mp)
for (j=0; j < dims[1]; j++) {
  for (ptrdiff_t i = 0; i < dims[0]; i++) {
  //...
  }
}
```

or use a `collapse(2)` directive to parallelize both the inner and outer loops:

```C
ptrdiff_t j, i;
#pragma omp parallel for collapse(2) if (use_mp)
for (j=0; j < dims[1]; j++) {
  for (i = 0; i < dims[0]; i++) {
  //...
  }
}
```

I would probably prefer the latter, because it offers more parallelism.

Reply:
Ah, that makes sense. I took `i` out of the loop by mistake. I didn't know about using `collapse()`, that seems very promising.